src/encoding/hybrid_rle/decoder.rs

```diff
@@ -3,38 +3,75 @@ use super::{super::ceil8, HybridEncoded};
 
 /// An iterator that, given a slice of bytes, returns `HybridEncoded`
 pub struct Decoder<'a> {
-    values: &'a [u8],
-    num_bits: u32,
+    inner: std::iter::Flatten<HybridIterator<'a>>,
 }
 
 impl<'a> Decoder<'a> {
-    pub fn new(values: &'a [u8], num_bits: u32) -> Self {
-        Self { values, num_bits }
+    pub fn new(values: &'a [u8], num_bits: u32, length: usize) -> Self {
+        Self {
+            inner: HybridIterator {
+                values,
+                num_bits,
+                remaining: length,
+            }
+            .flatten(),
+        }
     }
 }
 
 impl<'a> Iterator for Decoder<'a> {
-    type Item = HybridEncoded<'a>;
+    type Item = u32;
 
     fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+pub struct HybridIterator<'a> {
+    values: &'a [u8],
+    num_bits: u32,
+    remaining: usize,
+}
+
+impl<'a> Iterator for HybridIterator<'a> {
+    type Item = HybridEncoded<'a>;
+
+    fn next(&mut self) -> Option<HybridEncoded<'a>> {
         if self.values.is_empty() {
             return None;
         }
         let (indicator, consumed) = uleb128::decode(self.values);
         self.values = &self.values[consumed..];
         if indicator & 1 == 1 {
             // is bitpacking
-            let bytes = (indicator as usize >> 1) * self.num_bits as usize;
-            let result = Some(HybridEncoded::Bitpacked(&self.values[..bytes]));
-            self.values = &self.values[bytes..];
+            let num_bits = self.num_bits as usize;
```

Comment:
These changes mostly correspond to what was previously in `rle_decode` in `levels.rs`